cql3/restrictions/restriction.hh

```diff
@@ -58,6 +61,59 @@ namespace restrictions {
 struct allow_local_index_tag {};
 using allow_local_index = bool_class<allow_local_index_tag>;
 
+/// Work-in-progress new restriction representation.  When complete, it will replace the old one.
+namespace wip {
+
+// The new representation exposes its member data publicly.  Operations on restrictions are
+// performed by free functions that take restrictions as parameters and use the visitor pattern to
+// specialize code for different kinds of restrictions.
+//
+// The most interesting class is wip::binary_operator below, which can represent both multi- and
+// single-column restrictions.  Instead of the old merge_with mechanism for combining restrictions,
+// we will simply add them into a conjunction expression that will be processed using visitors.
+
+// Forward declare these, so std::variant can take them as arguments:
+struct binary_operator;
+struct conjunction;
+struct subscript;
+struct token;
+
+/// A restriction expression -- union of all possible restriction types.
+using expression = std::variant<binary_operator, conjunction>;
+
+/// Identifier(s) on LHS of an operator restriction.
+using idents = std::vector<const column_definition*>;
+
+/// LHS of an operator restriction.
+using lhs_t = std::variant<idents, token, subscript>;
+
+/// Represents token function on LHS of an operator relation.  No need to list column definitions
+/// here -- token takes exactly the partition key as its argument.
+struct token {};
+
+/// Subscripted column(s) on LHS of an operator restriction.  Eg: `m[123] > 456`.  Able to represent
+/// multiple columns, eg, `(x[1], y[2]) = (3, 4)`.
+struct subscript {
+    idents columns;
+    std::vector<::shared_ptr<term>> subs; ///< Corresponding subscripts.
+};
```

Comment:
Another approach is to have a tuple expression separate from a single-element subscript expression, and compose them. Of course, let's start with the easier path.

Reply:
I've since realized this and adjusted the code accordingly.  Stay tuned for the next version.

Reply:
Next version landed now...